charlesm93/math
# Issue 26 ode tuning

#### Submisison Checklist

- [X] Run unit tests: `./runTests.py test/unit/torsten`

#### Summary:
Solves https://github.com/charlesm93/stan/issues/6. Allows users to adjust the tuning parameters of the ODE integrators for generalOdeModel_*. There was technically nothing wrong with the functions in the math repo, but there was a disagreement between the math and the stan repos.

This pull request, combined with one I'm about to make on the Stan repo, fixes this issue by adding the `msgs` argument. I will eventually remove it because it doesn't get used. That said, it's harmless and it's a straightforward fix. The grammar files for generalOdeModel_* follow closely the architecture used for the integrate_ode_* functions, which have a msgs argument. This argument is treated differently than the other arguments, which is why I missed (removing) it when constructing the new grammar files; loosely speaking, it became part of the grammar for generalOdeModel_*. Changing that grammar requires a bit of refactoring and is not my immediate priority.

This fix is not ideal; we could attempt a better one, but the only effect would be a marginal improvement in code readability.

#### How to Verify:
This is tricky. In `integrate_ode_*.hpp` add a print statement (or an EXPECT_EQ statement) to see the value of the tuning parameters (rel_tol, abs_tol, and max_num_steps). 

I'm not sure how to design a unit test for this, because the error only comes up when I run a Stan model. I do not want to add an EXPECT_EQ statement inside a stan function (integrate_ode_* in this case). Doing so requires changing the Stan makefile to include the gtest library and means we'll have printing statements from the unit test while running Stan! Highly undesirable. 

Suggestions welcomed!

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Metrum Research Group, LLC

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
